beacon-chain/rpc/eth/rewards/handlers.go

```diff
@@ -166,22 +175,285 @@ func (s *Server) BlockRewards(w http.ResponseWriter, r *http.Request) {
+		return nil, false
+	}
+	st, err := s.Stater.StateBySlot(r.Context(), nextEpochEnd)
+	if err != nil {
+		errJson := &network.DefaultErrorJson{
+			Message: "Could not get state for epoch's starting slot: " + err.Error(),
+			Code:    http.StatusInternalServerError,
+		}
+		network.WriteError(w, errJson)
+		return nil, false
+	}
+	return st, true
+}
+
+func attRewardsBalancesAndVals(
+	w http.ResponseWriter,
+	r *http.Request,
+	st state.BeaconState,
+) (*precompute.Balance, []*precompute.Validator, []primitives.ValidatorIndex, bool) {
+	allVals, bal, err := altair.InitializePrecomputeValidators(r.Context(), st)
+	if err != nil {
+		errJson := &network.DefaultErrorJson{
+			Message: "Could not initialize precompute validators: " + err.Error(),
+			Code:    http.StatusBadRequest,
+		}
+		network.WriteError(w, errJson)
+		return nil, nil, nil, false
+	}
+	allVals, bal, err = altair.ProcessEpochParticipation(r.Context(), st, bal, allVals)
+	if err != nil {
+		errJson := &network.DefaultErrorJson{
+			Message: "Could not process epoch participation: " + err.Error(),
+			Code:    http.StatusBadRequest,
+		}
+		network.WriteError(w, errJson)
+		return nil, nil, nil, false
+	}
+	var rawValIds []string
+	if r.Body != http.NoBody {
+		if err = json.NewDecoder(r.Body).Decode(&rawValIds); err != nil {
+			errJson := &network.DefaultErrorJson{
+				Message: "Could not decode validators: " + err.Error(),
+				Code:    http.StatusBadRequest,
+			}
+			network.WriteError(w, errJson)
+			return nil, nil, nil, false
+		}
+	}
+	valIndices := make([]primitives.ValidatorIndex, len(rawValIds))
+	for i, v := range rawValIds {
+		index, err := strconv.ParseUint(v, 10, 64)
+		if err != nil {
+			pubkey, err := bytesutil.FromHexString(v)
+			if err != nil || len(pubkey) != fieldparams.BLSPubkeyLength {
+				errJson := &network.DefaultErrorJson{
+					Message: fmt.Sprintf("%s is not a validator index or pubkey", v),
+					Code:    http.StatusBadRequest,
+				}
+				network.WriteError(w, errJson)
+				return nil, nil, nil, false
+			}
+			var ok bool
+			valIndices[i], ok = st.ValidatorIndexByPubkey(bytesutil.ToBytes48(pubkey))
+			if !ok {
+				errJson := &network.DefaultErrorJson{
+					Message: fmt.Sprintf("No validator index found for pubkey %#x", pubkey),
+					Code:    http.StatusBadRequest,
+				}
+				network.WriteError(w, errJson)
+				return nil, nil, nil, false
+			}
+		} else {
+			if index >= uint64(st.NumValidators()) {
+				errJson := &network.DefaultErrorJson{
+					Message: fmt.Sprintf("Validator index %d is too large. Maximum allowed index is %d", index, st.NumValidators()-1),
+					Code:    http.StatusBadRequest,
+				}
+				network.WriteError(w, errJson)
+				return nil, nil, nil, false
+			}
+			valIndices[i] = primitives.ValidatorIndex(index)
+		}
+	}
+	if len(valIndices) == 0 {
+		valIndices = make([]primitives.ValidatorIndex, len(allVals))
+		for i := 0; i < len(allVals); i++ {
+			valIndices[i] = primitives.ValidatorIndex(i)
+		}
+	}
```

Comment:
isn't this being the default a little bit of a DoS vector?

Reply:
I don't think DoS-ing is that much of a concern in Beacon APIs, but you may want to raise an issue in the spec repo: https://github.com/ethereum/beacon-APIs